File: llm_generated_code/ChatGPT/Python/Level1/Q10.py

from collections import deque

def min_minutes_to_decay(grid):
    rows, cols = len(grid), len(grid[0])
    queue = deque()
    fresh_count = 0  # Count of healthy elements
    minutes = 0

    # Initialize the queue with all decayed elements and count healthy elements
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == 2:
                queue.append((r, c))  # Add all initially decayed elements
            elif grid[r][c] == 1:
                fresh_count += 1  # Count all healthy elements

    # Directions array for moving in 4 possible directions
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

    # BFS to spread decay
    while queue and fresh_count > 0:
        # Process all elements in the current level
        for _ in range(len(queue)):
            r, c = queue.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                # If the adjacent cell is healthy, decay it and add it to the queue
                if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == 1:
                    grid[nr][nc] = 2  # Decay the element
                    fresh_count -= 1
                    queue.append((nr, nc))
        minutes += 1

    # If there are any remaining healthy elements, return -1, otherwise return minutes
    return minutes if fresh_count == 0 else -1

File: llm_generated_code/ChatGPT/CPP/Level1/Q10.cpp

#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct Cell {
    int row, col, time;
};

int minDecayTime(vector<vector<int>>& grid) {
    int m = grid.size();
    int n = grid[0].size();
    queue<Cell> q;
    int healthyCount = 0;
    
    // Initialize the queue with all initially decayed elements and count healthy elements
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (grid[i][j] == 2) {
                q.push({i, j, 0});
            } else if (grid[i][j] == 1) {
                healthyCount++;
            }
        }
    }

    // Directions for moving up, down, left, right
    vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    int minutes = 0;

    // Perform BFS
    while (!q.empty()) {
        Cell cell = q.front();
        q.pop();
        int row = cell.row;
        int col = cell.col;
        int time = cell.time;
        minutes = max(minutes, time);

        // Spread decay to adjacent cells
        for (auto& dir : directions) {
            int newRow = row + dir.first;
            int newCol = col + dir.second;

            // Check if the new cell is within bounds and is healthy
            if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n && grid[newRow][newCol] == 1) {
                grid[newRow][newCol] = 2; // Decay the healthy element
                q.push({newRow, newCol, time + 1});
                healthyCount--; // Decrease the count of healthy elements
            }
        }
    }

    // If there are any healthy elements left, return -1, otherwise return the minutes
    return (healthyCount == 0) ? minutes : -1;
}

int main() {
    vector<vector<int>> grid = {
        {2, 1, 0, 2},
        {1, 0, 1, 2},
        {1, 1, 1, 0}
    };
    
    int result = minDecayTime(grid);
    if (result == -1) {
        cout << "It is not possible to decay all healthy elements." << endl;
    } else {
        cout << "Minimum number of minutes required to decay all healthy elements: " << result << endl;
    }
    
    return 0;
}
